### DISCUSSION TITLE
Using Spice TLines in RF analysis of a BFP193W amplifier

### DISCUSSION THREAD

**Discussion by tomhajjar** [ra3xdh/qucs_s#601] [Ideas]
A Spice lossless TLine can be used to emulate ideal microstrip lines. I converted Vadim's 2016 RF amplifier example to use TLines instead of Qucs Microstrip lines for the impedance matching networks. #94

I didn't have the original schematic so I assumed F=400 MHz. Using Transcalc I converted substrate parameters and dimensions of the microstrip lines to Z0 and Angle@400MHz. I then calculated Td. 

Qucs-S is generating the wrong netlist for a lossless TLine using Z0, F and NL so I had to use Z0 and Td. I made a TLine subcircuit and used both  the internal library BFP193W and the manufacturers BFP193W parasitic model.

Also attached is some testing of TLines.

![amp-ms](https://github.com/ra3xdh/qucs_s/assets/102838604/f43e6900-3ffc-42ca-8966-1e202ed21b82)

![2024-02-20_144832](https://github.com/ra3xdh/qucs_s/assets/102838604/d4ea1cf7-6f98-40e8-a469-31907834fdc4)

![2024-02-20_141218](https://github.com/ra3xdh/qucs_s/assets/102838604/d570b3ff-cec8-4a95-94a6-cf84a82358d6)

![2024-02-20_143835](https://github.com/ra3xdh/qucs_s/assets/102838604/6bdad5cc-fd4a-41a9-a1d1-89a2cf9aaa59)

![2024-02-20_141042](https://github.com/ra3xdh/qucs_s/assets/102838604/4d547332-7db5-4952-9dba-ef4e9ba53e54)

[BFP193W_amplifier_prj.zip](https://github.com/ra3xdh/qucs_s/files/14350366/BFP193W_amplifier_prj.zip)

[Transmission_Line_prj.zip](https://github.com/ra3xdh/qucs_s/files/14352092/Transmission_Line_prj.zip)

**Comment by ra3xdh** [comment]
Yes, simple MS line could be wrapped in LTRA or Tline model. It's need to provide a way to calculate the SPICE model parameters automatically. It still be unclear how to implement coupled microstrip.

**Comment by tomhajjar** [comment]
Coupled lines can be modeled like the lossless TLine using sources and resistors. There are many models available on the web and LTspice. Attached is one I found plus some LTspice examples. The other option is lumped models but they are narrower band.

How Qucs-S generates the TLine netlist has to be fixed. There are basically two modes, Z0/Td and Z0/F/NL. Qucs is adding Z0/Td/F/NL...etc to the netlist confusing ngspice which only wants to see Z0/Td or Z0/F/(NL). NL is assumed to be 0.25 if only Z0/F is in the netlist. 

[ICoupler.zip](https://github.com/ra3xdh/qucs_s/files/14359480/ICoupler.zip)

[Transmission Line.zip](https://github.com/ra3xdh/qucs_s/files/14364649/Transmission.Line.zip)

**Comment by dwarning** [comment]
Independent from the mentioned modes should ngspice make the job to set parameters if they are not given in Qucs-s interface.
See my patch #602

**Comment by tomhajjar** [comment]
I don't know what ngspice will do if the minimum parameters Z0/Td or Z0/F/(NL) are missing but if you put "conflicting" parameters in the netlist, instead of ngspice flagging an error it could generate erroneous data like attached. I think the netlist had the line "Td= ".

According to Holger, if you specify Td+F or Td+NL ngspice will choose only Td even if it is "Td=0" or "Td= ". You could ask Holger in link below.

Since Qucs-S is creating bad netlists, I would have to hand edit them and use ngspice in terminal mode to see what happens. 

https://sourceforge.net/p/ngspice/discussion/120973/thread/9bbd3f59c7/?limit=25#5432

![2024-02-20_192449](https://github.com/ra3xdh/qucs_s/assets/102838604/c072da83-49ff-4ddb-90c8-9481b5ad8e97)

**Comment by dwarning** [comment]
No errors! ngspice has defaults. But basically qucs-s can't feed ngspice with empty parameter strings. We speak about different things.
That your example with my patch:
![Bildschirmfoto vom 2024-02-21 22-04-52](https://github.com/ra3xdh/qucs_s/assets/18441932/6ce64acb-00fc-4516-b0be-3607254d9079)

**Comment by tomhajjar** [comment]
Unless someone makes me a Windows "nightly" release, I'm many weeks behind.

You could use this schematic version so Z, Td, F, NL are "visible". Confirmed using "Td= " causes the erroneous data. Td must default to a large number if "Td= ".

![2024-02-21_161712](https://github.com/ra3xdh/qucs_s/assets/102838604/ef90e93a-5f37-4fe6-8562-ea3dad9cf220)

![2024-02-21_162107](https://github.com/ra3xdh/qucs_s/assets/102838604/5d12069c-dccd-48ba-980e-0bf335a8c73d)

**Comment by tomhajjar** [comment]
@dwarning 

Since "Td=0" is valid entry and "Td= " is a possible entry, what entries for F or F/NL will "disable" Td and only use F or F/NL?

**Comment by dwarning** [comment]
"Td=0" and "Td=" are misleading to switch on delay specification by F and Nl.
So qucs-s netlister should only write Td if Td not equal Nl/F. Then the user choose specification by Td. Any other case is specification by F and Nl. Could it be the solution?

**Comment by tomhajjar** [comment]
Every entry variation needs to be tested to confirm the netlist is correct.

Users should not be allowed to enter both Td and F/NL. 

Another option is to have a check for Z0/Td and one for Z0/F/(NL). NL defaults to 0.25 if only F

**Comment by tomhajjar** [comment]
Until lossless TLine is fixed in Qucs-S I will work on other Spice compatible transmission line components.

Attached is an ideal lossless directional coupler provided by Intusoft decades ago. The model is 50 Ohms but can be scaled to any impedance. Caveats are the coupler has 0 insertion/coupling loss and will generate either very small or very large numbers if terminated in exactly 50 Ohms. Qucs/ngspice could generate errors using dB. I tested the model in Qucs-S and LTspice.

![2024-02-22_144458](https://github.com/ra3xdh/qucs_s/assets/102838604/caf29b87-9d04-4937-b702-54da85066b96)
![2024-02-22_150131](https://github.com/ra3xdh/qucs_s/assets/102838604/ed8234a9-8cef-45bb-8b91-047adeb0fbb2)
![2024-02-22_150154](https://github.com/ra3xdh/qucs_s/assets/102838604/45094493-056e-4527-8b0f-25725ff5f2b2)
![2024-02-22_150047](https://github.com/ra3xdh/qucs_s/assets/102838604/dce9b502-adcc-487d-90eb-7a2600c9f7e8)

[Directional Coupler.zip](https://github.com/ra3xdh/qucs_s/files/14378944/Directional.Coupler.zip)
